### PR TITLE
More features for C-Lisp

### PR DESCRIPTION
**Pull request by GlowingScrewdriver**
Implemented the following:
* `while` loop
* Block scopes
* 1-D Arrays

### CODE REVIEW

**Comment by GlowingScrewdriver** [comment]
Scopes are introduced by block statements, functions, `if`/`else`, `for`, and `while` -- in general, any collection of statements.

File: src/backend/c-lisp.py [hunk 1]
```diff
@@ -14,7 +14,9 @@ class BrilispCodeGenerator:
     def __init__(self):
         # Type tracking
         self.symbol_types = {}  # Variable name -> type
+        self.scopes = []  # Stack of scope tags
         self.function_types = {}  # Function name > (ret-type, (arg-types...))
+        self.pointer_types = {}  # For internal use, e.g. temporary pointer variables
```
Comment (chsasank) [inline src/backend/c-lisp.py:19]: why can we not use symbol types for this?

File: src/backend/c-lisp.py [hunk 2]
```diff
@@ -284,6 +404,14 @@ def gen_expr(self, expr, res_sym=None):
             return self.gen_var_expr(expr, res_sym)
         elif self.is_binary_expr(expr):
             return self.gen_binary_expr(expr, res_sym)
+        elif self.is_ptradd_expr(expr):
```
Comment (chsasank) [inline src/backend/c-lisp.py:407]: is ptradd etc too different from fadd etc instructions?

Comment (GlowingScrewdriver) [reply src/backend/c-lisp.py:407]: The main difference is this: the other expressions ( e.g. `fadd`, `add`) have produces values of a fixed type (`float` and `int` in this example). For this reason, even temporary symbols created by `ptradd` instructions need to be tracked, unlike with the others.

For example, in the expression `(load (ptradd arr 5))`,
* `ptradd` produces a temporary symbol
* `load` emits a Brilisp `load` that loads from the temporary symbol. The type of the resulting value, the value that is being loaded, cannot be determined unless the temporary symbol from `ptradd` is tracked.

**Comment by GlowingScrewdriver** [comment]
Added more array tests.
The `matrix-sum` test uses arrays allocated using both `alloc` expression (stack allocation) and standard C library's `malloc` function (heap allocation)!

**Comment by GlowingScrewdriver** [comment]
Added tests `gaussian-elimination` and `insertion-sort`.
These two tests take inputs from the command line, using C standard library functions `atoi` and `strtof`

**Comment by GlowingScrewdriver** [comment]
Will fix #34  and #33

File: src/backend/c-lisp.py (hunk 1 above)
Comment (chsasank) [reply src/backend/c-lisp.py:19]: Because we used typed pointers, we need to have this?

File: src/backend/c-lisp.py [hunk 3]
```diff
@@ -42,6 +44,9 @@ def __init__(self):
             "fgt": "bool",
             "fle": "bool",
             "fge": "bool",
+            # Boolean logic
+            "and": "bool",
```
Comment (chsasank) [inline src/backend/c-lisp.py:48]: add "not"

File: src/backend/c-lisp.py [hunk 4]
```diff
@@ -51,6 +56,13 @@ def c_lisp(self, prog):
 
         return ["brilisp"] + [self.gen_function(fn) for fn in prog[1:]]
 
+    def get_scoped_name(self, name):
```
Comment (chsasank) [inline src/backend/c-lisp.py:59]: use scope0.scope1..scopen.name

Comment (chsasank) [reply src/backend/c-lisp.py:59]: same goes for brilisp random variables

File: src/backend/c-lisp.py [hunk 5]
```diff
@@ -59,18 +71,21 @@ def gen_function(self, func):
...
                 raise CodegenError(f"Bad function prototype: {func[1]}")
 
-        self.symbol_types = {}  # Clear the symbol table
+        # Clear the symbol table and scope stack
+        self.symbol_types = {}
+        self.scopes = [random_label(CLISP_PREFIX, ["scope"])]
+
         name, ret_type = func[1][0]
         parm_types = []
+        header = func[1][0:1]
```
Comment (chsasank) [inline src/backend/c-lisp.py:80]: [func[1]]

File: src/backend/c-lisp.py [hunk 6]
```diff
@@ -59,18 +71,21 @@ def gen_function(self, func):
             if not len(elem) == 2:
                 raise CodegenError(f"Bad function prototype: {func[1]}")
 
-        self.symbol_types = {}  # Clear the symbol table
+        # Clear the symbol table and scope stack
+        self.symbol_types = {}
+        self.scopes = [random_label(CLISP_PREFIX, ["scope"])]
```
Comment (chsasank) [inline src/backend/c-lisp.py:76]: fix this based on above common

Comment (chsasank) [reply src/backend/c-lisp.py:76]: In fact scopes should be []

File: src/backend/c-lisp.py [hunk 7]
```diff
@@ -59,18 +71,21 @@ def gen_function(self, func):
...
+
         name, ret_type = func[1][0]
         parm_types = []
+        header = func[1][0:1]
         for parm in func[1][1:]:
-            parm_types.append(parm[1])
-            self.symbol_types[parm[0]] = parm[1]
+            typ = parm[1]
```
Comment (chsasank) [inline src/backend/c-lisp.py:82]: and this function doesn't have to be modified

File: src/backend/c-lisp.py [hunk 8]
```diff
@@ -95,6 +112,35 @@ def gen_stmt(self, stmt):
...
 
+    def is_while_stmt(self, stmt):
+        return stmt[0] == "while"
+
+    def gen_while_stmt(self, stmt):
+        if len(stmt) < 3:
+            raise CodegenError(f"Bad while statement: {stmt}")
+
+        cond_sym, loop_lbl, cont_lbl, break_lbl = [
+            random_label(CLISP_PREFIX, [extra])
```
Comment (chsasank) [inline src/backend/c-lisp.py:123]: use scopes for all intermediate variables

**Comment by chsasank** [comment]
Scope code can be better structured. Added comments.

File: src/backend/c-lisp.py (hunk 8 above)
Comment (GlowingScrewdriver) [reply src/backend/c-lisp.py]: Done

File: src/backend/c-lisp.py (hunk 7 above)
Comment (GlowingScrewdriver) [reply src/backend/c-lisp.py]: Done

File: src/backend/c-lisp.py (hunk 5 above)
Comment (GlowingScrewdriver) [reply src/backend/c-lisp.py]: Done

File: src/backend/c-lisp.py (hunk 4 above)
Comment (GlowingScrewdriver) [reply src/backend/c-lisp.py:59]: Done

File: src/backend/c-lisp.py (hunk 3 above)
Comment (GlowingScrewdriver) [reply src/backend/c-lisp.py]: Done

File: src/backend/c-lisp.py (hunk 8 above)
Comment (chsasank) [reply src/backend/c-lisp.py:129]: decided against it after looking at the outputs

File: src/backend/c-lisp.py [hunk 9]
```diff
@@ -248,28 +309,96 @@ def is_var_expr(self, expr):
         return isinstance(expr, str)
 
     def gen_var_expr(self, expr, res_sym):
-        if expr in self.symbol_types:
-            return [["set", [res_sym, self.symbol_types[expr]], ["id", expr]]]
+        scoped_name = self.scoped_lookup(expr)
+        if scoped_name in self.symbol_types:
+            typ = self.symbol_types[scoped_name]
+            instr_list = [["set", [res_sym, typ], ["id", scoped_name]]]
+            if typ[0] == "ptr":
```
Comment (chsasank) [inline src/backend/c-lisp.py:316]: create is_pointer(typ) function

File: src/backend/c-lisp.py [hunk 10]
```diff
@@ -248,28 +309,96 @@ def is_var_expr(self, expr):
...
+        return expr[0] == "ptradd"
+
+    def gen_ptradd_expr(self, expr, res_sym):
+        if len(expr) != 3:
+            raise CodegenError(f"Bad ptradd expression: {expr}")
+
+        offset_sym = random_label(CLISP_PREFIX)
+        ptr_name = self.scoped_lookup(expr[1])
+        ptr_type = self.symbol_types[ptr_name]
+        self.pointer_types[res_sym] = ptr_type
```
Comment (chsasank) [inline src/backend/c-lisp.py:352]: store all intermediate symbol types. I don't want pointers to be very special.

File: src/backend/c-lisp.py (hunk 2 above)
Comment (chsasank) [reply src/backend/c-lisp.py:416]: This is also why we need self.pointer_types

File: src/backend/c-lisp.py (hunk 10 above)
Comment (chsasank) [reply src/backend/c-lisp.py:352]: For now passing this. We'll think about it later

**Pull request merged**